feat: Send GraphQL "operationName" in HTTP breadcrumbs

maxchuquimia opened the pull request:
## :scroll: Description

This PR attempts to support sending GraphQL operation names with existing HTTP breadcrumbs.

## :bulb: Motivation and Context

When using GraphQL requests, Sentry's HTTP breadcrumbs are not valuable as they all show the same URL. Being able to see the operation name in a HTTP breadcrumb is extremely useful when tracking down what happened leading up to an error.

The main reason I am opening this PR now is because of Sentry's move to distributing XCFrameworks. This has made my fork unusable as pointing our app's `Package.swift` to our fork doesn't immediately change the downloaded `.binaryTarget` in Sentry's `Package.swift`. At a glance, the GitHub Actions in my fork are failing so I thought opening a PR may be the easiest option to enable us to update Sentry (which we now need to do to get `PrivacyInfo.xcprivacy`).

Relates to https://github.com/getsentry/sentry-cocoa/issues/3494

## :green_heart: How did you test it?

- Our original fork https://github.com/maxchuquimia/sentry-cocoa/commit/fa0b6607ed04ca3af9f3cabe588f13c319a24e25 has been in production for a number of months with no issues.
- Added some unit tests before opening this PR

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps
- Perhaps there's a better way?
- Ideally someone else tests this manually
- Unit tests could probably be expanded or improved on, they were difficult to jump in and write as I have done no other work in this project 😅

brustolin commented on Sources/Sentry/include/NSURLSessionTask+Sentry.m:
```diff
@@ -0,0 +1,25 @@
+#import "NSURLSessionTask+Sentry.h"
+
+
+@implementation
+NSURLSessionTask (Sentry)
+
+- (nullable NSString *)sentry_graphQLOperationName
+{
+    if (!self.originalRequest.HTTPBody) { return nil; }
+    if (![[self.originalRequest valueForHTTPHeaderField:@"Content-Type"] isEqual: @"application/json"]) { return nil; }
+
+    NSError *error = nil;
+    id requestDictionary = [NSJSONSerialization JSONObjectWithData:self.originalRequest.HTTPBody options:0 error:&error];
+
+    if (error) { return nil; }
+    if (![requestDictionary isKindOfClass: [NSDictionary class]]) { return nil; } // Could be an array
+
+    id operationName = [requestDictionary valueForKey:@"operationName"];
+    if (![operationName isKindOfClass: [NSString class]]) { return nil; }
+    if ([operationName length] == 0) { return nil; }
+
+    return operationName;
+}
```
We can't use Objc Categories to extend a class because our static lib will strip it. We need to turn this into a C function.

maxchuquimia replied:
Cheers @brustolin, I couldn't see an example of how this has been done before so I have changed the function to be like other ones in nearby files. If this still won't work, please let me know where I can find an example of a C-style helper function!

brustolin replied:
This works fine. Thanks!!

brustolin reviewed (commented):
Hello @maxchuquimia. Thank you so much for the help. 
The PR looks very good, with tests and all.

I left a comment of something that needs changing.
The PR is also missing a CHANGELOG entry.

Nice work!!

maxchuquimia commented:
> I left a comment of something that needs changing. The PR is also missing a CHANGELOG entry.

@brustolin updated!

brustolin commented:
This looks good to me. 
I dont see a reason to not have this feature merged.

@philipphofmann, @kahest thoughts?

philipphofmann commented on Sources/Sentry/Public/SentryOptions.h:
```diff
@@ -142,6 +142,12 @@ NS_SWIFT_NAME(Options)
+ * Whether to attach the top level `operationName` node of HTTP json requests to HTTP breadcrumbs
+ * @note Default is @c NO.
+ */
+@property (nonatomic, assign) BOOL enableGraphQLOperationTracking;
```
`m`: Please add some tests for this new property. All you have to do is to add a similar test as this one https://github.com/getsentry/sentry-cocoa/blob/32ac9341b30033766e7ff03cd2e29592ad90e6ff/Tests/SentryTests/SentryOptionsTest.m#L198-L201

philipphofmann commented on Sources/Sentry/include/NSURLSessionTask+Sentry.m:
```diff
@@ -0,0 +1,24 @@
+
+@implementation SentryNSURLSessionTask
+
++ (nullable NSString *)graphQLOperationNameFromTask:(NSURLSessionTask *_Nonnull)task
```
`m`: It would be great if we could write this code in Swift. See https://github.com/getsentry/sentry-cocoa/blob/main/develop-docs/Swift-Usage.md for some guidance.

philipphofmann replied:
`m`: It would be great to have some unit tests to cover all the edge cases in this class.

philipphofmann reviewed (commented):
Thanks a lot for this PR. I think we only need a couple of improvements then we can merge this. Most important we should align with how we add GraphQL breadcrumbs in our [Java SDK](https://github.com/getsentry/sentry-java/blob/8d236f4091e1c1ed890ebfd78e58f6746487cd0e/sentry/src/main/java/io/sentry/Breadcrumb.java#L193-L223), see also our [develop docs](https://develop.sentry.dev/sdk/features/#graphql-client-integrations). If you can't easily retrieve `operation_name, `operation_type` and `operation_id`, it's also acceptable to only do one of them.

maxchuquimia commented:
Hey @philipphofmann, cheers for your comments! I have made the following changes:
- Added missing options test
- Rewrote operation name extraction in Swift (which looks fine to me but it does mean I can no longer say I have been using the exact code in this PR in production!)
- Add further tests for operation name extraction
- Renamed `graphql` to `operation_name` in breadcrumb (I'm not at all sure how to get the `operation_id`, and the only way I could see to get `operation_type` was from a proprietary Apollo header so I left it out for now)

philipphofmann commented on Sources/Sentry/SentryNetworkTracker.m:
```diff
@@ -489,6 +504,10 @@ - (void)addBreadcrumbForSessionTask:(NSURLSessionTask *)sessionTask
             [NSHTTPURLResponse localizedStringForStatusCode:responseStatusCode];
+
+        if (self.isGraphQLOperationTrackingEnabled) {
+            breadcrumbData[@"operation_name"] = [sessionTask getGraphQLOperationName];
```
`m`: On Java, they have more information, and they even add an extra breadcrumb for GraphQL. Let's prefix this with for now until we can provide more information.
```suggestion
            breadcrumbData[@"graphql_operation_name"] = [sessionTask getGraphQLOperationName];
```

philipphofmann commented on Sources/Sentry/SentryNetworkTracker.m:
```diff
@@ -440,6 +450,11 @@ - (void)captureFailedRequests:(NSURLSessionTask *)sessionTask
     context[@"response"] = response;
+
+    if (self.isGraphQLOperationTrackingEnabled) {
+        context[@"operation_name"] = [sessionTask getGraphQLOperationName];
```
`m`: On Java, they have more information for the context.. Let's prefix this with for now until we can provide more information.
```suggestion
        context[@"graphql_operation_name"] = [sessionTask getGraphQLOperationName];
```

philipphofmann reviewed (commented):
Two minor things to fix. Then we can merge this. Thank you @maxchuquimia 👍

maxchuquimia commented:
@philipphofmann Done! 🎉

philipphofmann reviewed (approved):
LGTM, thank you @maxchuquimia 🥇 .

philipphofmann reviewed (approved):
LGTM

maxchuquimia commented:
Hey @brustolin, you mentioned "We can't use Objc Categories to extend a class" -- as part of later comments I was asked to migrate the code in this PR to Swift, does this rule apply to Swift Extensions somehow? Now that this has been merged we are attempting to use it but are receiving this crash, I'm not sure how the function could not be there seeing as it compiled just fine 😅 

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFLocalDataTask getGraphQLOperationName]: unrecognized selector sent to instance 0x1047251b0'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000180491128 __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x000000018008412c objc_exception_throw + 56
	2   CoreFoundation                      0x00000001804a5f78 +[NSObject(NSObject) instanceMethodSignatureForSelector:] + 0
	3   CoreFoundation                      0x0000000180495278 ___forwarding___ + 1280
	4   CoreFoundation                      0x000000018049759c _CF_forwarding_prep_0 + 92
	5   App                                 0x00000001106cfff8 -[SentryNetworkTracker addBreadcrumbForSessionTask:] + 732
	6   App                                 0x00000001106cee50 -[SentryNetworkTracker urlSessionTask:setState:] + 516
	7   App                                 0x00000001106acf5c __57+[SentryNetworkTrackingIntegration swizzleURLSessionTask]_block_invoke_2.33 + 64
	8   CFNetwork                           0x00000001842724a4 _CFHTTPMessageSetResponseProxyURL + 6408
	9   CFNetwork                           0x0000000184270e34 _CFHTTPMessageSetResponseProxyURL + 664
	10  CFNetwork                           0x0000000184272d20 _CFHTTPMessageSetResponseProxyURL + 8580
	11  CFNetwork                           0x00000001842e1f38 CFURLDownloadCancel + 44640
	12  CFNetwork                           0x00000001842e0288 CFURLDownloadCancel + 37296
	13  CFNetwork                           0x00000001842df820 CFURLDownloadCancel + 34632
	14  CFNetwork                           0x00000001843cadf4 _CFNetworkSetATSContext + 79752
	15  CFNetwork                           0x00000001842e2f88 CFURLDownloadCancel + 48816
	16  libdispatch.dylib                   0x0000000103d9a038 _dispatch_block_async_invoke2 + 104
	17  libdispatch.dylib                   0x0000000103d8993c _dispatch_client_callout + 16
	18  libdispatch.dylib                   0x0000000103d91bd8 _dispatch_lane_serial_drain + 916
	19  libdispatch.dylib                   0x0000000103d9291c _dispatch_lane_invoke + 420
	20  libdispatch.dylib                   0x0000000103d9f2f8 _dispatch_root_queue_drain_deferred_wlh + 324
	21  libdispatch.dylib                   0x0000000103d9e754 _dispatch_workloop_worker_thread + 488
	22  libsystem_pthread.dylib             0x0000000103ce3814 _pthread_wqthread + 284
	23  libsystem_pthread.dylib             0x0000000103ce25d4 start_wqthread + 8
)
libc++abi: terminating due to uncaught exception of type NSException
```

philipphofmann commented:
Hmm, it could be that Swift extension methods don't work. @maxchuquimia, can you try to convert the getGraphQLOperationName to a helper function that takes `URLSessionTask` as an argument and see if it solves the issue? Then, we could open a quick PR to address the issue.

maxchuquimia commented:
PR opened @philipphofmann - https://github.com/getsentry/sentry-cocoa/pull/3973
I can't test if it solves the issue due to the original xcframework issue...
